include/SDL3/SDL_ioasync.h

```diff
@@ -0,0 +1,500 @@
+ * synchronous way, but it can allow for more efficiency, and never having
+ * framerate drops as the hard drive catches up, etc.
+ *
+ * The general usage pattern for async i/o is:
+ *
+ * - Create one or more SDL_IOAsyncGroup.
+ * - Open files with SDL_IOAsyncFromFile.
+ * - Start i/o tasks to the files with SDL_ReadIOAsync or SDL_WriteIOAsync,
+ *   putting those tasks into one of the groups.
+ * - Later on, use SDL_GetIOAsyncResult on a group to see if any task
+ *   is finished without blocking. Tasks might finish in any order with
+ *   success or failure.
+ * - When all your tasks are done, close the file with SDL_CloseIOAsync.
+ *   This also generates a task, since it might flush data to disk!
+ *
+ * This all works, without blocking, in a single thread, but one can also use
+ * a task group in a background thread and sleep until new results have
+ * arrived:
+ *
+ * - Call SDL_WaitIOAsyncResult from one or more threads to efficiently block
+ *   until new tasks complete.
+ * - When shutting down, call SDL_SignalIOAsyncGroup to unblock any sleeping
+ *   threads despite there being no new tasks completed.
+ *
+ * And, of course, to match the synchronous SDL_LoadFile, we offer
+ * SDL_LoadFileAsync as a convenience function. This will handle allocating
+ * a buffer, slurping in the file data, and null-terminating it; you still
+ * get a task handle to check later or add to a task group.
+ */
+
+#ifndef SDL_ioasync_h_
+#define SDL_ioasync_h_
+
+#include <SDL3/SDL_stdinc.h>
+
+#include <SDL3/SDL_begin_code.h>
+/* Set up for C function definitions, even when using C++ */
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * The asynchronous I/O operation structure.
+ *
+ * This operates as an opaque handle. One can then request read or write
+ * operations on it.
+ *
+ * \since This struct is available since SDL 3.0.0.
+ *
+ * \sa SDL_IOAsyncFromFile
+ */
+typedef struct SDL_IOAsync SDL_IOAsync;
+
+/**
+ * Types of asynchronous i/o tasks.
+ *
+ * \since This enum is available since SDL 3.0.0.
+ */
+typedef enum SDL_IOAsyncTaskType
+{
+    SDL_IOASYNC_TASK_READ,   /**< A read operation. */
+    SDL_IOASYNC_TASK_WRITE,  /**< A write operation. */
+    SDL_IOASYNC_TASK_CLOSE   /**< A close operation. */
+} SDL_IOAsyncTaskType;
+
+/**
+ * Possible outcomes of an asynchronous i/o task.
+ *
+ * \since This enum is available since SDL 3.0.0.
+ */
+typedef enum SDL_IOAsyncResult
+{
+    SDL_IOASYNC_COMPLETE,  /**< request was completed without error */
+    SDL_IOASYNC_FAILURE,   /**< request failed for some reason; check SDL_GetError()! */
+    SDL_IOASYNC_CANCELLED  /**< request was cancelled before completing. */
+} SDL_IOAsyncResult;
+
+/**
+ * Information about a completed asynchronous i/o request.
+ *
+ * \since This struct is available since SDL 3.0.0.
+ */
+typedef struct SDL_IOAsyncOutcome
+{
+    SDL_IOAsync *ioasync;   /**< what generated this task. This pointer will be invalid if it was closed! */
```

Comment:
```suggestion
    SDL_IOAsync *ioasync;   /**< what generated this task. This pointer will be NULL if it was closed! */
```

Reply:
It won't be NULL, it'll be the pointer that was valid before closing the file invalidated it...but in case the app is keying off it, we give them the pointer.